### PR TITLE
Update QuickFixFinder UI and Refactor Filters for Reusability

### LINKED ISSUE #238: Update QuickFixFinder UI to add list of workers and filter on search query entered

### PR DESCRIPTION
**Pull request by Autek**
### Description:

This PR updates the user interface of `QuickFixFinder.kt` and, as part of the process, refactors the filtering logic into a dedicated, reusable component.

Key changes include:

- **UI Updates in QuickFixFinder**:
  - Remove button with `Search` and `Announce` when a search query is entered in search bar
  - add the filtering component when a search query is entered in search bar

- **Refactoring of Filter Logic**:
  - Moved all filter-related operations (`reapplyFilters`, `clearFilters`, etc.) and UI elements (`SearchFilterButtons`, `FilterRow`) from `SearchWorkerResult.kt` into a new `SearchFilters.kt` component.

- **Separation of Concerns**:
  - `QuickFixFinder.kt` now focuses on the UI layout and user flow.
  - Filtering logic and UI components are modularized, making them easier to maintain and reuse.

---

### Benefits:
1. Improved and more intuitive UI within `QuickFixFinder.kt`.
2. Cleaner separation of concerns between core filtering logic and the main UI code.
3. Easier future maintenance and enhancements due to reusable filtering components.

---

### Test Plan:
- Verify that the updated UI in `QuickFixFinder.kt` displays as intended.
- Confirm that all filters (Availability, Service Type, Price Range, Location, and Highest Rating) still work correctly after the refactoring.
- Ensure the "Clear" button resets all filters and returns the full list of profiles.
- Check that filter icon colors and transitions behave as expected.

---

### Screenshots:
Capture d’écran 2024-12-14 à 20 58 56
Capture d’écran 2024-12-14 à 20 58 41

### CODE REVIEW

**Comment by Autek** [comment]
This PR is almost ready for review! 
I just need to do some final code cleanup before marking it as ready. Should be good to go very soon.  
Thank you for your patience!

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt
Comment (ramshty) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt:1]: good refactor to make use of call backs to be able to use the profile selected and display it in the bottom sheet. There is also no absolute value and everything is relative. Good job !

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchFilters.kt [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+          leadingIcon = Icons.Default.WorkspacePremium,
+          trailingIcon = if (ratingFilterApplied) Icons.Default.Clear else null,
+          applied = ratingFilterApplied),
+      SearchFilterButtons(
+          onClick = { onShowPriceRangeBottomSheet() },
+          text = "Price Range",
+          leadingIcon = Icons.Default.MonetizationOn,
+          trailingIcon = Icons.Default.KeyboardArrowDown,
+          applied = priceFilterApplied))
+}
```
Comment (ramshty) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchFilters.kt:179]: really good way to modularize the code. Nice job!

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchFilters.kt
Comment (ramshty) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchFilters.kt:1]: This file is the proof of your effort to avoid duplication of code and your desire to keep the code clean and concise. Thank you for the refactor and the modularization.

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt
Comment (ramshty) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt:1]: the logic is good and is exactly what you discussed. Good job !

**Review by ramshty — DISMISSED** [review]
Overall really good work and nice effort of modularization. Your code is very clean and concise, i doubt that we could do better. Thanks for your work !!

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt:1]: You have absolute padding values in your code. You need to refactor this file to support relative positioning.

The composable is already wrapped within a `BoxWithConstraints`, you can use it ;)

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt [hunk 2]
```diff
@@ -153,8 +212,9 @@ fun QuickFixScreenTab(
             color =
                 if (pagerState.currentPage == currentPage) colorScheme.background
                 else colorScheme.tertiaryContainer,
-            style = MaterialTheme.typography.titleMedium,
+            style = typography.titleMedium,
             modifier =
-                Modifier.padding(horizontal = 16.dp, vertical = 8.dp).testTag("tabText$title"))
+                Modifier.padding(horizontal = screenWidth * 0.04f, vertical = screenWidth * 0.02f)
+                    .testTag("tabText$title"))
```
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/QuickFixFinder.kt:218]: Nice! thanks for refactoring this 😄

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt [hunk 3]
```diff
@@ -1,212 +1,394 @@
...
+    }
+  }
   BoxWithConstraints {
     val widthRatio = maxWidth.value / 411f
     val heightRatio = maxHeight.value / 860f
     val sizeRatio = minOf(widthRatio, heightRatio)
-    val screenHeight = maxHeight.value
-    val screenWidth = maxWidth.value
+    val screenHeight = maxHeight
+    val screenWidth = maxWidth
```
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt:181]: These changes remove a fix to a buggy representation of the search bar on smaller phones. Drop these changes please.¹

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt [hunk 4]
```diff
@@ -1,212 +1,394 @@
...
+                              searchQuery = it
+                              searchViewModel.searchEngine(it, profiles as List<WorkerProfile>)
+                            },
+                            shape = CircleShape,
+                            textStyle = poppinsTypography.bodyMedium,
+                            textColor = colorScheme.onBackground,
+                            placeHolderColor = colorScheme.onBackground,
+                            leadIconColor = colorScheme.onBackground,
+                            widthField = 300.dp * widthRatio,
+                            heightField = 40.dp * heightRatio,
```
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchOnBoarding.kt:225]: linked to ¹

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt [hunk 5]
```diff
@@ -92,65 +55,51 @@ import com.arygm.quickfix.model.quickfix.QuickFixViewModel
...
-  var baseLocation by remember { mutableStateOf(phoneLocation) }
+  val (uiState, setUiState) = remember { mutableStateOf(SearchUIState()) }
+  var isWindowVisible by remember { mutableStateOf(false) }
   val context = LocalContext.current
+  val locationHelper = LocationHelper(context, MainActivity())
+  var selectedWorkerProfile by remember { mutableStateOf(WorkerProfile()) }
+  val filterState = rememberSearchFiltersState()
+  val defaultBitmap =
+      Bitmap.createBitmap(100, 100, Bitmap.Config.ARGB_8888) // Example default Bitmap
+
```
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt:85]: Same comment here for the `defaultBitmap` value

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt [hunk 6]
```diff
@@ -92,65 +55,51 @@ import com.arygm.quickfix.model.quickfix.QuickFixViewModel
...
     if (locationHelper.checkPermissions()) {
       locationHelper.getCurrentLocation { location ->
         if (location != null) {
-          phoneLocation =
-              com.arygm.quickfix.model.locations.Location(
-                  location.latitude, location.longitude, "Phone Location")
-          baseLocation = phoneLocation
+          val userLoc = Location(location.latitude, location.longitude, "Phone Location")
+          filterState.phoneLocation = userLoc
         } else {
```
Comment (AnotherMedo) [inline app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt:103]: Why the change ?

**Review by AnotherMedo — CHANGES_REQUESTED** [review]
Quite a bit of changes to keep the code maintainable, consistent and bug free. But otherwise great PR! 

@ramshty will review the functionality of your branch by checking out to it

File: app/src/androidTest/java/com/arygm/quickfix/ui/search/ProfileResultsTest.kt [hunk 7]
```diff
@@ -1,148 +1,116 @@
...
-                reviews = ArrayDeque(),
-                location = null,
-                price = 59.0,
-            ))
-
-    // Mock the AccountViewModel to return sample account data
+    // Mock AccountViewModel fetchUserAccount
     every { accountViewModel.fetchUserAccount(any(), captureLambda()) } answers
         {
           val uid = firstArg<String>()
```
Comment (AnotherMedo) [inline app/src/androidTest/java/com/arygm/quickfix/ui/search/ProfileResultsTest.kt:40]: Why was this test removed ?

File: app/src/androidTest/java/com/arygm/quickfix/ui/search/SearchWorkerResultScreenTest.kt [hunk 8]
```diff
@@ -212,35 +213,15 @@ class SearchWorkerResultScreenTest {
...
-          quickFixViewModel,
-          workerViewModel = workerViewModel)
-    }
-    // Set the search query and verify that the title and description match the query
-    searchViewModel.setSearchQuery("Unknown")
-
-    // Check if the description with the query text is displayed
-    composeTestRule.onAllNodesWithText("Unknown").assertCountEquals(3)
-  }
-
```
Comment (AnotherMedo) [inline app/src/androidTest/java/com/arygm/quickfix/ui/search/SearchWorkerResultScreenTest.kt:243]: Again, why was this test removed ? Once added and passed, tests should in no way be removed. Please restore it

**Review by AnotherMedo — CHANGES_REQUESTED** [review]
⚠️ This comment is exclusively for the **testing** part of your Pull Request.

Please make sure to rollback the changes on tests if you want this PR merged.

I have noticed that you made the test files more trivial and I do not understand why... See the comments for more details.

**Review by Sebbayra — CHANGES_REQUESTED** [review]
I checked out on the branch to check if the filters still work, however there are some functionalities that have been changed and don't work, like for example we can't stack the rating filter on other filters, the emergency one bugs as well and doesn't stack, the location filter doesn't change colors when applied, and the normal search doesn't display results, and I have seen that you have changed the tests for SearchWorkerResult have been changed and I really want their logic and functionality unchanged because they are crucial to ensure that every filter works as intended.

File: app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt (hunk 6 above)
Comment (Autek) [reply app/src/main/java/com/arygm/quickfix/ui/uiMode/appContentUI/userModeUI/search/SearchWorkerResult.kt:103]: because I found it more intuitive to have the phone location in the filterState but I can change it if you feel that it is a bad idea :)

File: app/src/androidTest/java/com/arygm/quickfix/ui/search/ProfileResultsTest.kt (hunk 7 above)
Comment (Autek) [reply app/src/androidTest/java/com/arygm/quickfix/ui/search/ProfileResultsTest.kt:40]: this test was not removed but updated to match the update of the `ProfileResults` composable. It was replaced by a new test that is testing exactly the same thing

File: app/src/androidTest/java/com/arygm/quickfix/ui/search/SearchWorkerResultScreenTest.kt (hunk 8 above)
Comment (Autek) [reply app/src/androidTest/java/com/arygm/quickfix/ui/search/SearchWorkerResultScreenTest.kt:243]: This test was removed because the `searchQuery` field in `searchViewModel` appears to be unused and serves no functional purpose in the codebase. As a result, the test doesn’t provide any meaningful validation. At least from my understanding

**Comment by Autek** [comment]
**Note:** This PR will not be merged as it is too late and there are still some unresolved bugs like @Sebbayra pointed out.
Requested changes are implemented though.

**Comment by AnotherMedo** [comment]
> **Note:** This PR will not be merged as it is too late and there are still some unresolved bugs like @Sebbayra pointed out. Requested changes are implemented though.

Closing this since it will not be merged as stated in the comment
